src/Featurizers/ForwardFillImputerFeaturizer.h

```diff
@@ -0,0 +1,185 @@
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         ForwardFillImputerTransformer
+///  \brief         Imputer that populates null incoming values with the last
+///                 non-null value encountered.
+///
+template <typename T>
+class ForwardFillImputerTransformer :
+    public Components::InferenceOnlyTransformerImpl<
+        typename Traits<T>::nullable_type,
+        T
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    static_assert(Traits<T>::IsNullableType == false || Traits<T>::IsNativeNullableType, "'T' must not be a nullable type");
+
+    using BaseType =
+        Components::InferenceOnlyTransformerImpl<
+            typename Traits<T>::nullable_type,
+            T
+        >;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    ForwardFillImputerTransformer(void) = default;
+    ForwardFillImputerTransformer(T defaultValue);
+    ForwardFillImputerTransformer(Archive &ar);
+
+    ~ForwardFillImputerTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(ForwardFillImputerTransformer);
+
+    bool operator==(ForwardFillImputerTransformer const &other) const;
+    bool operator!=(ForwardFillImputerTransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    T                                       _lastValue;
+    bool                                    _validLastValue = false;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        // ----------------------------------------------------------------------
+        using TheseTraits                   = Traits<typename BaseType::InputType>;
+        // ----------------------------------------------------------------------
+
+        if(TheseTraits::IsNull(input)) {
+            if(_validLastValue == false) {
+                // This scenario will happen when the first imputed item is null, as we haven't seen a valid value yet.
+                throw std::runtime_error("No source value for forward fill");

```

Comment:
Is there any way to suppress this error and let it continue? This is one of the use cases AutoML told us is the reasons that they will use multiple imputation strategies, so that if any of them still have nulls that another imputation strategy will catch it.

Reply:
There is, but the code path should only be exercised if the output type is optional and (possibly) they opt into the behavior. I'm going to add a TODO comment for now, but I want to spend some time thinking about the resulting complexity before implementing the functionality.

---
In reply to: [383480925](https://github.com/microsoft/FeaturizersLibrary/pull/144#discussion_r383480925) [](ancestors = 383480925)